Support multiple redirect_uris when creating OAuth 2.0 Applications

ThisIsMissEm opened the pull request:
This brings us more inline with the dynamic client registration specification, where you can create a single application with multiple redirect URIs, see the example request at: https://datatracker.ietf.org/doc/html/rfc7591#section-3.1

This works around some weirdness in doorkeeper, where it only has a `redirect_uri` field on an Application, but it does some weird "join multiple values with a \n for the database" thing.

Arguably, this could be considered a breaking change because it introduces some kinda weird behavior, but it is technically the "correct" behavior with regards to OAuth 2.0...

This is currently a draft as it requires test coverage to be written, and I have had the time to write said tests yet.

ThisIsMissEm committed:
Support multiple redirect_uris when creating OAuth 2.0 Applications

This brings us more inline with the dynamic client registration specification

ThisIsMissEm commented on app/serializers/rest/application_serializer.rb:
```diff
@@ -19,6 +19,18 @@ def client_secret
     object.secret
   end
 
+  def redirect_uri
+    # Doorkeeper stores the redirect_uri value as a newline delimeted list in
+    # the database, as we may have more than one redirect URI, return the first:
+    object.redirect_uri.split.first
+  end
```
This method / property should probably be considered deprecated, because previously it *could* return a value that included a new line character, which wouldn't be a valid URI, if the application was created not through the API.

i.e., if you go to the Development > Applications section in the settings, then the redirect_uris is a textarea allowing newline characters; It just so happens that we've no way to request that application back through the API to hit the previously faulty code here.

ThisIsMissEm replied:
@ClearlyClaire up to you as to what we do with the behaviour here: do we do a "correct" behaviour, or do we retain the previous behaviour?

ClearlyClaire replied:
I *think* I would keep the existing behavior, however weird it is.

ThisIsMissEm replied:
Have retained that behaviour in the latest commit.

ThisIsMissEm commented on app/controllers/api/v1/apps_controller.rb:
```diff
@@ -24,6 +24,6 @@ def app_scopes_or_default
 
   def app_params
-    params.permit(:client_name, :redirect_uris, :scopes, :website)
+    params.permit(:client_name, :scopes, :website, :redirect_uris, redirect_uris: [])
```
This actually took me surprisingly long to figure out that I could write it like this.

ThisIsMissEm replied:
I think we may actually want this to be with `require` instead for several of these. e.g., 
```suggestion
    params.require(:client_name, :redirect_uris, redirect_uris: []).permit(:scopes, :website)
```

i.e., you definitely can't create an application without a `client_name` and `redirect_uris` value.

But `scopes` and `website` are actually optional.

ClearlyClaire replied:
Confusingly, this is not how `require` works, see https://api.rubyonrails.org/classes/ActionController/Parameters.html#method-i-require

ThisIsMissEm replied:
Hmm, okay, well, we may want to think about some better input validation here, since there is potential for this to be an attack surface.

ThisIsMissEm replied:
I'm not sure; just feels like we should be doing more validation here, rather than relying on the Model validations to catch issues.

ThisIsMissEm committed:
Improve specs for /api/v1/apps

ThisIsMissEm commented on spec/requests/api/v1/apps_spec.rb:
```diff
@@ -9,7 +9,7 @@
 
     let(:client_name)   { 'Test app' }
-    let(:scopes)        { nil }
+    let(:scopes)        { 'read write' }
```
Previously we were falling back to default scopes, when the primary way of registering an oauth app is with scopes.

ThisIsMissEm commented on spec/requests/api/v1/apps_spec.rb:
```diff
@@ -87,8 +125,67 @@
-    context 'without required params' do
-      let(:client_name)   { '' }
+    context 'with multiple redirect_uris as a string' do
+      let(:redirect_uris) { "https://redirect1.example/\napp://redirect2.example/" }
```
This is previously accepted string input, and would result in multiple redirect_uri's being stored / usable.

ThisIsMissEm commented:
The following issue was discovered whilst working on the tests: https://github.com/mastodon/mastodon/issues/30152

ClearlyClaire commented on app/serializers/rest/application_serializer.rb:
```diff
@@ -19,6 +19,18 @@ def client_secret
+  def redirect_uris
+    # Doorkeeper stores the redirect_uri value as a newline delimeted list in
+    # the database:
+    object.redirect_uri.split
```
I'd suggest splitting the same way doorkeeper internally does, which is omitting the `split` parameter, so the current version is fine.

ThisIsMissEm commented on app/serializers/rest/application_serializer.rb:
```diff
@@ -19,6 +19,19 @@ def client_secret
+  def redirect_uri
+    # Doorkeeper stores the redirect_uri value as a newline delimeted list in
+    # the database, as we may have more than one redirect URI, return the first:
+    object.redirect_uri.split.first
```
```suggestion
    # This returns potentially multiple redirect URIs as a newline delimetered string
    object.redirect_uri
```

ThisIsMissEm committed:
Retain existing weird behaviour regarding multiple redirect_uri values

ThisIsMissEm commented:
Working on a few additional changes after noticing some gaps when authoring the documentation for this PR.

ThisIsMissEm committed:
Expose redirect_uris via /api/v1/apps/verify_credentials

ThisIsMissEm committed:
Add spec to ensure /api/v1/apps/verify_credentials never exposes the client_secret

ThisIsMissEm commented:
@ClearlyClaire okay, we're good — I've added:
* exposing the `redirect_uris` property from `/api/v1/apps/verify_credentials` (we had started to expose `client_id` and `scopes` in #27142 
* added test case to ensure `/api/v1/apps/verify_credentials` never exposes `client_secret` (just in case the controller is changed in the future)
* Moved the `redirect_uris` method from the serializer to the `Doorkeeper::Application` extension (I feel like this is cleaner)

ThisIsMissEm commented on app/controllers/api/v1/apps/credentials_controller.rb:
```diff
@@ -4,6 +4,6 @@ class Api::V1::Apps::CredentialsController < Api::BaseController
     return doorkeeper_render_error unless valid_doorkeeper_token?
 
-    render json: doorkeeper_token.application, serializer: REST::ApplicationSerializer, fields: %i(name website vapid_key client_id scopes)
+    render json: doorkeeper_token.application, serializer: REST::ApplicationSerializer, fields: %i(name website vapid_key client_id scopes redirect_uris)
```
@ClearlyClaire is there a way to write this to "exclude" certain fields? Because that's what we really want here, instead of "include" certain fields. (we're excluding `client_secret`)

ClearlyClaire replied:
I don't think so. At least I haven't found one.

ClearlyClaire replied:
I think the current version is good enough, honestly. Though that suggestion isn't too bad either.

ThisIsMissEm replied:
I've done that split now since it was easy to do, we now have `Application` for verify and `CredentialApplication` for registration.

ThisIsMissEm commented on app/serializers/rest/application_serializer.rb:
```diff
@@ -19,6 +19,9 @@ def client_secret
   end
 
+  # We should consider this property deprecated for 4.3.0
+  delegate :redirect_uri, to: :object
```
huh, maybe? yeah... let me test and adjust if necessary.

ThisIsMissEm committed:
Refactor to use attribute instead of delegate for :redirect_uri

ThisIsMissEm committed:
Refactor to split ApplicationSerializer into CredentialApplicationSerializer for /api/v1/apps and ApplicationSerializer for /api/v1/apps/verify_credentials

ThisIsMissEm commented on app/serializers/rest/application_serializer.rb:
```diff
@@ -1,12 +1,15 @@
-  attributes :id, :name, :website, :scopes, :redirect_uri,
-             :client_id, :client_secret
+  attributes :id, :name, :website, :scopes, :redirect_uris,
+             :client_id
```
I had to retain `:client_id` here, since it [was previously exposed](https://github.com/mastodon/mastodon/commit/7c3fea727583dd9bf482729cc87272e976a25b2d) on `/api/v1/apps/verify_credentials`, even though it's not strictly necessary for it to be exposed there. (The addition of `client_id` to `ApplicationSerializer` has gone out in a release potentially, even though it appears to only be on `main`

ThisIsMissEm replied:
Actually, I've removed `client_id` here, since that commit was never in a tagged release from what I can tell.

ClearlyClaire replied:
What was the rationale for including `client_id` previously, and why did you decide to stop exposing it?

ThisIsMissEm committed:
Refactor to remove client_id from ApplicationSerializer since this was never released

ThisIsMissEm commented on app/lib/application_extension.rb:
```diff
@@ -23,6 +23,12 @@ def confirmation_redirect_uri
     redirect_uri.lines.first.strip
   end
```
@ClearlyClaire I'm not sure what this property is, but it looks odd. Should probably be `redirect_uri.split.first`

ThisIsMissEm commented on app/serializers/rest/application_serializer.rb:
```diff
@@ -1,24 +1,18 @@
-
-  def client_secret
-    object.secret
-  end
```
In 4.2.x, the properties of `Application` for `client_id` and `client_secret` were optional; it's only in `main` that `client_id` started to be always returned.

ClearlyClaire replied:
Calling them optional is confusing.

If my understanding is correct:
- before #27142, `client_id` and `client_secret` were only present in the response to `POST /api/v1/apps`
- after #27142, `client_secret` is still only present in the response to `POST /api/v1/apps` but `client_id` is exposed in response to `GET /api/v1/apps/verify_credentials` as well

And your change would revert to before #27142.

ThisIsMissEm replied:
yup, which I think is fine & intentional, since #27142 was never released. This makes scopes and redirect_uris always returned inline with that PR. I don't know if you need `client_id` in verify?

ThisIsMissEm replied:
we can moved it (`client_id`) back in I guess? I can't recall why I added it to the verify response

ClearlyClaire reviewed (approved):
Looks good to me although I'm confused why this is going back on `client_id` being introspectable.
